uaclient/util.py

```diff
@@ -293,12 +293,18 @@ def get_dict_deltas(
 
 
 @lru_cache(maxsize=None)
-def get_machine_id(data_dir: str) -> str:
+def get_machine_id(cfg) -> str:
     """Get system's unique machine-id or create our own in data_dir."""
-    # Generate, cache our own uuid if not present on the system
+    # Generate, cache our own uuid if not present in config or on the system
     # Docker images do not define ETC_MACHINE_ID or DBUS_MACHINE_ID on trusty
     # per Issue: #489
-    fallback_machine_id_file = os.path.join(data_dir, "machine-id")
+
+    if cfg.machine_token:
+        cfg_machine_id = cfg.machine_token.get("machineId")
+        if cfg_machine_id:
+            return cfg_machine_id
+
+    fallback_machine_id_file = os.path.join(cfg.data_dir, "machine-id")
 
     for path in [ETC_MACHINE_ID, DBUS_MACHINE_ID, fallback_machine_id_file]:
```

Comment:
What do you think if we reorder this check so that we source fallback_machine_id_file first if present. The only way it would be present is if we have written it already either due to falling through to the uuid4 generation below on first call or if we add a `self.cfg.write_cache("machine-id", response.get("machineId") if "machineId" in machine_token` over in [UAContractClient.request_contract_machine_attach](https://github.com/canonical/ubuntu-advantage-client/blob/main/uaclient/contract.py#L88) and [UAContractClient._request_machine_token_update](https://github.com/canonical/ubuntu-advantage-client/blob/main/uaclient/contract.py#L221)

Reply:
In side discussion @renanrodrigo pointed out we can't prefer the fallback_machine_id_file above ETC and DBUS sources because those will change across cloned images on some clouds and we don't have visibliity to that change if we continue to prefer our now stale <data_path>/machine-id cache. So we can't reorder.

We decided as well that we need to clear the lru_cache for `util.get_machine_id.clear_cache()` in UAContractClient whenever a new machine_token response is persisted if the 'machineId' key exists in the machine-token and the token has changed. This avoids the cache on get_machine_id returning the previous machine_id which was pulled from ETC or DBUS.